aten/src/ATen/native/cuda/Copy.cu

```diff
@@ -265,6 +265,12 @@ static bool copy_requires_temporaries(TensorIterator& iter, bool p2p_enabled) {
   }
 
   bool same_dtype = iter.dtype(0) == iter.dtype(1);
+
+  // Check if the tensor is 1D or 2D and non-contiguous
+  if (iter.ndim() <= 2 && !iter.is_contiguous()) {  
```

Comment:
you also should check for `same_dtype` here? Please add corresponding test case.

Reply:
temporaries are requires if there's a non-contiguous iterator of dim2 (the condition above is satisfied) but dtypes are different. You would already return `false` in this case.